File: crates/nw-gamedata-codegen/resources/go/standalone/game-assets/filesystem.go
package gameassets

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

func LoadLooseDatasheets(root string) ([]DatasheetAsset, error) {
	resolvedRoot, err := canonicalPath(root)
	if err != nil {
		return nil, fmt.Errorf("resolve datasheet root %s: %w", root, err)
	}
	var assets []DatasheetAsset
	if err := collectLooseDatasheets(resolvedRoot, resolvedRoot, &assets); err != nil {
		return nil, err
	}
	return assets, nil
}

func collectLooseDatasheets(root string, dir string, assets *[]DatasheetAsset) error {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return err
	}
	sort.Slice(entries, func(i int, j int) bool { return entries[i].Name() < entries[j].Name() })
	for _, entry := range entries {
		path := filepath.Join(dir, entry.Name())
		if entry.IsDir() {
			if err := collectLooseDatasheets(root, path, assets); err != nil {
				return err
			}
			continue
		}
		if !entry.Type().IsRegular() || !IsDatasheetPath(path) {
			continue
		}
		bytes, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		if !IsDatasheetBytes(bytes) {
			continue
		}
		relative, err := filepath.Rel(root, path)
		if err != nil {
			return err
		}
		if relativePathEscapesRoot(relative) {
			return fmt.Errorf("datasheet path %s is outside root %s", path, root)
		}
		*assets = append(*assets, DatasheetAsset{
			Path:  NormalizeVirtualPath(relative),
			Bytes: bytes,
		})
	}
	return nil
}

File: crates/nw-gamedata-codegen/resources/typescript/standalone/game-assets/filesystem.ts
import { readdir, readFile, realpath } from "node:fs/promises";
import { isAbsolute, join, relative, sep } from "node:path";

import { normalizeVirtualPath } from "./catalog.js";
import { type DatasheetAsset, isDatasheetBytes, isDatasheetPath } from "./datasheet.js";

export async function loadLooseDatasheets(root: string): Promise<DatasheetAsset[]> {
  const resolvedRoot = await realpath(root);
  const assets: DatasheetAsset[] = [];
  await collectLooseDatasheets(resolvedRoot, resolvedRoot, assets);
  return assets;
}

async function collectLooseDatasheets(root: string, dir: string, assets: DatasheetAsset[]): Promise<void> {
  const entries = await readdir(dir, { withFileTypes: true });
  entries.sort((left, right) => left.name.localeCompare(right.name));
  for (const entry of entries) {
    const path = join(dir, entry.name);
    if (entry.isDirectory()) {
      await collectLooseDatasheets(root, path, assets);
      continue;
    }
    if (!entry.isFile() || !isDatasheetPath(path)) {
      continue;
    }
    const bytes = await readFile(path);
    if (!isDatasheetBytes(bytes)) {
      continue;
    }
    const relativePath = relative(root, path);
    if (isOutsideRelativePath(relativePath)) {
      throw new Error(`datasheet path ${path} is outside root ${root}`);
    }
    assets.push({
      path: normalizeVirtualPath(relativePath),
      bytes
    });
  }
}

function isOutsideRelativePath(path: string): boolean {
  return path === ".." || path.startsWith(`..${sep}`) || isAbsolute(path);
}
